### DISCUSSION TITLE
Accessing typegen information attached to an xstate-machine in library code

### DISCUSSION THREAD

**Discussion by UberMouse** [statelyai/xstate#3552] [Ideas] [answered]
I write a library built on top of xstate, xstate-tree, for binding react views to xstate machines. As part of this library there is an abstraction layer between the machine and the view so the view does not have any direct reference to the state machine. This is done by creating `selectors` that take in `context` and `inState` and the return value is passed to the view, same concept for `actions` which provide callbacks to attach to UI events which in turn send actions to the machine.

Both the `selector` function and the `view` are passed an `inState` function, which is `(state: UnionOfMachineStates) => boolean` to determine if a machine is in a specific state. Up until recently the typings for this function had used type states to extract the union of states. But since we have recently started using typegen and typestates are deprecated, I recently landed a change that uses the `matchesStates` property of the typegen information to type the state union if the machine is using typegen, falling back to the typestate information for legacy machines.

This was the solution I came up for accessing that type information
```
function buildSelectors<
  TContext,
  TStateSchema,
  TEvent extends EventObject,
  TTypestate extends Typestate<TContext>,
  TXstateActions extends BaseActionObject,
  TServices extends ServiceMap,
  TTypegen extends TypegenConstraint,
  TSelectors,
  TStates = TTypegen extends TypegenEnabled
    ? TTypegen extends ResolveTypegenMeta<infer T, any, any, any>
      ? T extends TypegenMeta
        ? T["matchesStates"]
        : never
      : never
    : TTypestate["value"]
>(
  __machine: StateMachine<
    TContext,
    TStateSchema,
    TEvent,
    TTypestate,
    TXstateActions,
    TServices,
    TTypegen
  >,
  selectors: Selectors<TContext, TEvent, TSelectors, TStates>
): (
  ctx: TContext,
  canHandleEvent: (e: TEvent) => boolean,
  inState: (state: TStates) => boolean,
  currentState: TStates
) => TSelectors
```

However @Andarist instructed me that might break at any time and to open an issue to discuss the use-case, so here I am.

**Comment by Andarist** [comment, marked as answer]
Could you showcase how this is used by the consumer? And what exact shape of `TStates` do you assume when you consider compound and parallel states?

**Reply by UberMouse** [reply]
TStates is the type of the arguement of inState, so `inState: (state: TStates) => boolean`. The consumer uses that in the React view `if (inState("loaded.doingAThing")) {}` etc. Compound states are.like.this, same as it worked for typestates. Parallel states don't come into the typing, just need validate the state is valid on the machine. 

But it's implemented in xstate-tree by passing it directly to a `matches` method on an xstate state object.

**Reply by Andarist** [reply]
Ok, so if this matches the `matches` directly 1-to-1 then I think using `StateFrom<MachineType>["matches"]` should be the way to go:
```ts
import { createMachine, StateFrom } from "xstate";

const m = createMachine({
  tsTypes: {} as { "@@xstate/typegen": true; matchesStates: "a" | "b" },
});

const myMatches: StateFrom<typeof m>["matches"] = m.initialState.matches;
```

[TS playground](https://www.typescriptlang.org/play?ts=4.7.4#code/JYWwDg9gTgLgBAbzgYygUwIYzQWQ8gC2ADs0AaOAZRizQDEoIQ4BfOAM0eYCIAPAZxrZuAbjgB6cXBgBPMGn4AuOABYAdAGYNalQChdyCMUFxmAXhTpaeQiTQAKBLrjT+AFTkLlCNhn6I4bgABIIEhNHFZeQBzNGJuZRgoAFc0MRAsQgVqWiVAjG44AB9AgCNCljJdFgBKEX1DY3gQGTwYLLyc7AYmAB4otAh2UwA+AG1uDPaCBW4AXTgLEDUSYBhgDAAbLrQ1KY76oA)

**Reply by UberMouse** [reply]
Did not know about `StateFrom`, I think I can change it so the `inState` function is typed as the `matches` property of a machines state instead of trying to extract the state.  But it doesn't work for reasons I don't quite understand 

[TS playground](https://www.typescriptlang.org/play?ts=4.7.4#code/JYWwDg9gTgLgBAbzgYygUwIYzQWQ8gC2ADs0AaOAZRizQDEoIQLra9CS04BfOAM0Yg4AIgAeAZxrZhAbjgB6eXBgBPMGnEAuOABYAdAGYDenQChTyCMUlwhAXhTo2+IqQAUCU3GXiAKmo1tBF4McUQRAAEIiSk0eVV1AHM0YmFtGCgAVzQ5ECxCDVZsLREMYTgAHxEAI3LuMlNuAEoZc0treBAVPBgCkqL6QQAeBLQIPlsAPgBtYTzegg1hAF04BxA9EmAYYAwAGwG9eb7W0z5M4mQdq1sMAGs0ADVgNAB3Ia84X3ZXLjRRbDEAAmYQGP04QwwxBUFChMLgcNh0KR8MRCOhkwa3l8AzCDgGDCYQ2+Lk4Mzm+UW4hWWK+AAVGGA8YgSANtG5JLRtDjYuImmtJnBqhAIHtMMRuKZJm48hxSNzwaQKAA3F6vdlgRklXwMiBM-l2QVwpoarXc3X6gXolSIT7oGCZKDEODETJ7PYIsJw1qSixWGyqt5rW4PZ5vGUUDys2LNK2ebzACZuaO0NzCMpNfmeT6S5qtIA)

**Reply by Andarist** [reply]
In here it’s best to use `TMatches` and not `TStates`:
[TS playground](https://www.typescriptlang.org/play?ts=4.7.4#code/JYWwDg9gTgLgBAbzgYygUwIYzQWQ8gC2ADs0AaOAZRizQDEoIQLra9CS04BfOAM0Yg4AIgAeAZxrZhAbjgB6eXBgBPMGnEAuOABYAdAGYDenQChTyCMUlwhAXhTo2+IqQAUCU3GXiAKmo1tBF4McUQRAAEIiSk0eVV1AHM0YmFtGCgAVzQ5ECxCDVZsLREMYTgAHxEAI3LuMlNuAEoZc0treBAVPBgCkqL6QQAeBLQIPlsAPgBtYTzegg1hAF04BxA9EmAYYAwAGwG9eb7W0z5M4mQdq1sMAGs0ADVgNAB3Ia84X3ZXLjRRbDEAAmYQGP04QwwxBUFChMLgcNh0KR8MRCOhkwa3m++UWYQcAwYTCGOI4pBmc1xS2WWK+AAVGGB8YgSANtDiFhpuKZJm48mS0OzwaQKAA3F6vbRuMCMkq+BkQJlNNaTdEqJpSmWKuUKpUqtWIT7oGCZKDEODETJ7PYIsJw1rcixWGzit5rW4PZ5vPkUDys2LNfWebzACZuf20NzCMpNZWeT7c5qtIA)

**Reply by UberMouse** [reply]
Ahhh very obvious mistake now that I see it. So we end up with [this](https://www.typescriptlang.org/play?ts=4.7.4#code/JYWwDg9gTgLgBAbzgYygUwIYzQWQ8gC2ADs0AaOAZRizQDEoIQKBBYgT2tr0JLTgC+cAGaMQcAEQAPAM41sEgNxwA9Crgx2YNDIBccACwA6AMwmjBgFCXkEYnLjiAvCnTd8RUgAoEluBpkAFS0dfQQhDBlESQABGNl5NBVNbQBzNGIJfRgoAFc0ZRAsQh0ubD1JDAk4AB9JACNqgTJLAQBKRWtbe3gQdjwYEoqy+jEAHhS0CGFHAD4AbQkiwYIdCQBdOBcQIxJgGGAMABsRo2WhzsthXOJkA7tHDABrNAA1YDQAdzG-OECeTz8NBSbDEAAmUTYnESAL4LX8-2KqyiLhGDCYY0RvFICyWSLW63hfwACowwCjECQRvpESsdAJLLMvEVsWgabDSBQAG4fT76LxgMkVQKkiDktpbWZwDAcNr8wVi4Wi8WS6UcRC-dAwXJQYhwYi5I5HaVRGXsToMmx2Bw8r5bR4vd5fZkUHxUxLtVW+fzAGZed20LwSKptCW+X4M9qdIA), very nice. I have some refactoring to do :)

Thanks for the help

**Reply by Andarist** [reply]
No problem! We should probably nest typegen stuff in some extra property like `__dontUseThisAskForHelp` 😝
